### PR TITLE
[SPARK-23591][SQL] Add interpreted execution to EncodeUsingSerializer

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

The PR adds interpreted execution to EncodeUsingSerializer.

## How was this patch tested?

added UT

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 1]
```diff
@@ -1154,8 +1154,18 @@ case class CreateExternalRow(children: Seq[Expression], schema: StructType)
...
   extends UnaryExpression with NonSQLExpression {
 
-  override def eval(input: InternalRow): Any =
-    throw new UnsupportedOperationException("Only code-generated evaluation is supported")
+  override def eval(input: InternalRow): Any = {
+    val obj = child.eval(input)
+    if (obj == null) return null
+
+    val conf = Option(SparkEnv.get).map(_.conf).getOrElse(new SparkConf)
+    val serializer = if (kryo) {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: Can you add this as a transient lazy field? That will save us from recreating serializer instances all over the place.

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: yes, actually the same could/should be done also for the code generated version, shouldn't it? If so, should I do both changes here or in another PR?

Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: If I am not mistaken the code generated version adds the serializer instance to classes immutable state, so there is not need.

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: Yes, you're right, sorry. I just realized after my comment. Sorry.

Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: No problem :)

**Comment by mgaido91** [comment]
retest this please

**Comment by mgaido91** [comment]
any more comments on this @hvanhovell ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 2]
```diff
@@ -105,6 +105,66 @@ trait InvokeLike extends Expression with NonSQLExpression {
...
+trait BaseSerializer {
+  /**
+   * If true, Kryo serialization is used, otherwise the Java one is used
+   */
+  val kryo: Boolean
+
+  /**
+   * The serializer instance to be used for serialization/deserialization
+   */
+  lazy val serializerInstance = {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: We can also put this in a companion and call this from whole stage code gen.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 3]
```diff
@@ -105,6 +105,66 @@ trait InvokeLike extends Expression with NonSQLExpression {
...
+        new JavaSerializer(conf)
+      }
+    s.newInstance()
+  }
+
+  /**
+   * The name of the variable referencing the serializer, which is added with
+   * `addImmutableSerializerIfNeeded`
+   */
+  lazy val serializerVarName = if (kryo) {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: Let's make `addImmutableSerializerIfNeeded` return the serializer name. No need for a separate method.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala (hunk 2 above)
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: The major concern I have on this is performance. Referencing an outer object means that is is put into an array and every time it is called we have an array reference call and a cast before its usage. I saw many concerns in previous PRs on codegen about doing these things in order not to affect performances.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 4]
```diff
@@ -105,6 +105,66 @@ trait InvokeLike extends Expression with NonSQLExpression {
   }
 }
 
+/**
+ * Common trait for [[DecodeUsingSerializer]] and [[EncodeUsingSerializer]]
+ */
+trait BaseSerializer {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: `BaseSerializer` is a bit generic. Can we come up with a better name? We could name it `UsingSerializer` or `SerializerSupport` (I am not super excited about either, so we could also leave it as it is).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala (hunk 2 above)
Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: I was talking about the code below (the comment was somewhat terse). We should definitely put the `serializerInstance` into a the generated classes immutable state. All I am saying is that is that we can put the factory code below in a companion, and use that in both the generated and interpreted code paths. This reduces the amount of code generated, and also makes things easier to test.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala (hunk 4 above)
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: I can't think of any better name honestly.... I'd go for `SerializerSupport`, thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala (hunk 2 above)
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: Oh, sorry, I see now what you mean. Yes, I will do that, thanks.

**Comment by hvanhovell** [comment]
LGTM - merging to master. Thanks!

**Pull request closed without merging**